### PR TITLE
add plain and cholesky rank updates

### PR DESCRIPTION
**Pull request by karlnapf**

### CODE REVIEW

**Comment by karlnapf** [comment]
fixes #4385 and prepares ground for a linear regression with bias update

**Comment by karlnapf** [comment]
Eigen3 can also do this, but it only updates the lower or upper part of a matrix. We might want to think about introducing types link symmetric matrices where only half of the memory is effectively used and the rest can be used a temp storage by algorithms, like eigen does

**Comment by karlnapf** [comment]
@OXPHOS btw is there a reason why we don't define the methods in linalg using a type trait 

```
template <typename T, typename T2 = typename std::enable_if_t<std::is_arithmetic<T>::value>>
template <typename T, typename T2 = typename std::enable_if_t<std::is_floating_point<T>::value>>
template <typename T, typename T2 = typename std::enable_if_t<std::is_integral<T>::value>>
```
etc

Could this replace the macros to define things for certain ptypes?

**Comment by OXPHOS** [comment]
I can not recall any reason for not using `enable_if`. We actually used it
for one scenario. I can look into it soon. Just finished my defense ;D

**Comment by karlnapf** [comment]
wow congrats! :)

File: src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp [hunk 1]
```diff
@@ -266,6 +291,25 @@ void LinalgBackendEigen::pinv_impl(
...
 
+template <typename T>
+void LinalgBackendEigen::rank_update_impl(
+    SGMatrix<T>& A, const SGVector<T>& b, T alpha) const
+{
+	T x;
+	T update;
+	for (auto j : range(A.num_cols))
+	{
+		x = b[j];
```
Comment (iglesias) [inline src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp]: Out of curiosity, having the local x is better than just using `b[j]` all over the method? Same question for `update`.

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 2]
```diff
@@ -680,9 +673,63 @@ namespace shogun
...
+			REQUIRE(
+			    L.num_rows == b.size(),
+			    "Vector size (%d) must match matrix size (%dx%d)\n", b.size(),
+			    L.num_rows, L.num_rows);
+			return infer_backend(L, SGMatrix<T>(b))
+			    ->cholesky_rank_update(L, b, alpha, lower);
+		}
+
+		/**
+		 * Updates a matrix \f$A\f$ with a rank one term in-place,If A = LL^T
```
Comment (iglesias) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: Minor: I think for this method the the A = LL^T can be left out.

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 3]
```diff
@@ -680,9 +673,63 @@ namespace shogun
...
+			    "Matrix dimensions (%dx%d) are not square\n", A.num_rows,
+			    A.num_cols);
 			return infer_backend(A)->cholesky_factor(A, lower);
 		}
 
+		/**
+		 * Updates the Cholesky factorization \f$A = L L^{*}\f$ with a rank one
+		 * term in-place.
+		 * If A = LL^T before the update, then after it
+		 * LL^{*} = A + alpha * b b^{*}
```
Comment (iglesias) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: Quite nitpicking: perhaps a different name for the L as they are not the same as the ones above. In wikipedia they use tilde analogously for A. It is completely subjective I think, whether we want the doc to be more math correct, or just reflect what happens in the code. I am happy either way :-)

**Pull request merged**